drv/sidecar-seq-server/src/main.rs

```diff
@@ -24,177 +29,335 @@ use i2c_config::devices;
+                return Ok(());
+            }
+        }
+
+        Err(SeqError::SequencerTimeout)
+    }
+
+    fn power_down(&mut self) -> Result<(), SeqError> {
+        ringbuf_entry!(Trace::InitiateTofinoPowerDown);
+        Ok(self
+            .sequencer
+            .set_enable(false)
+            .map_err(|_| SeqError::SequencerError)?)
     }
 
-    fn led_toggle(&mut self) {
-        if self.led_on {
-            self.led_off();
-        } else {
-            self.led_on();
+    fn handle_tick(&mut self) -> Result<(), SeqError> {
+        let state = self.sequencer.state()?;
+        let error = self.sequencer.error()?;
+
+        ringbuf_entry!(Trace::TofinoSequencerTick(self.policy, state, error));
+
+        match (self.policy, state, error) {
+            // Power down if the Tofino should be disabled.
+            (TofinoSequencerPolicy::Disabled, TofinoSeqState::InPowerUp, _) => {
+                self.power_down()
+            }
+            (TofinoSequencerPolicy::Disabled, TofinoSeqState::A0, _) => {
+                self.power_down()
+            }
+            // Power up
+            (
+                TofinoSequencerPolicy::LatchOffOnFault,
+                TofinoSeqState::A2,
+                TofinoSeqError::None,
+            ) => self.power_up(),
+            _ => Ok(()), // Do nothing by default.
         }
     }
 }
 
+struct ServerImpl {
+    mainboard_controller: MainboardController,
+    clock_generator: ClockGenerator,
+    tofino: Tofino,
+}
+
 impl idl::InOrderSequencerImpl for ServerImpl {
-    fn get_state(
+    fn tofino_seq_policy(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<TofinoSequencerPolicy, RequestError<SeqError>> {
+        Ok(self.tofino.policy)
+    }
+
+    fn set_tofino_seq_policy(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        policy: TofinoSequencerPolicy,
+    ) -> Result<(), RequestError<SeqError>> {
+        ringbuf_entry!(Trace::TofinoSequencerPolicyUpdate(policy));
+        self.tofino.policy = policy;
+        Ok(())
+    }
+
+    fn tofino_seq_state(
         &mut self,
         _: &RecvMessage,
-    ) -> Result<PowerState, RequestError<SeqError>> {
-        ringbuf_entry!(Trace::GetState);
-        Ok(self.state)
+    ) -> Result<TofinoSeqState, RequestError<SeqError>> {
+        Ok(self.tofino.sequencer.state().map_err(SeqError::from)?)
     }
 
-    fn set_state(
+    fn tofino_seq_error(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<TofinoSeqError, RequestError<SeqError>> {
+        Ok(self.tofino.sequencer.error().map_err(SeqError::from)?)
+    }
+
+    fn clear_tofino_seq_error(
         &mut self,
         _: &RecvMessage,
-        state: PowerState,
     ) -> Result<(), RequestError<SeqError>> {
-        ringbuf_entry!(Trace::SetState(self.state, state));
+        ringbuf_entry!(Trace::ClearingTofinoSequencerFault(
+            self.tofino.sequencer.error().map_err(SeqError::from)?
```

Comment:
Do we want to return early if logging `error()` fails, or just if `clear_error` fails below?